File: ptolemy/media/Audio.java
/* A library of audio operations.

Copyright (c) 1998-2000 The Regents of the University of California.
All rights reserved.

Permission is hereby granted, without written agreement and without
license or royalty fees, to use, copy, modify, and distribute this
software and its documentation for any purpose, provided that the above
copyright notice and the following two paragraphs appear in all copies
of this software.

IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
ENHANCEMENTS, OR MODIFICATIONS.

                                        PT_COPYRIGHT_VERSION_2
                                        COPYRIGHTENDKEY

@ProposedRating Red (eal@eecs.berkeley.edu)
@AcceptedRating Red (cxh@eecs.berkeley.edu)
*/

package ptolemy.media;

import java.lang.*;
import java.util.*;
import java.io.*;


//////////////////////////////////////////////////////////////////////////
//// Audio
/**
 * Instances of this class represent audio data equivalent to that
 * contained by a Sun/NeXT audio file (.au file).  The class also
 * includes a set of utility static methods for manipulating audio
 * signals.
 * Currently, only an 8kHz sample rate, mu-law encoded, monophonic
 * audio format is supported.
 * <p>
 *  The format of an audio file is:
 *  <CENTER>
 *  <TABLE BORDER = 1>
 *  <TR BGCOLOR = #DFDFA0><TD>byte</TD><TD>type</TD>
 *  <TD>field name   </TD><TD> field value                      </TD></TR>
 *  <TR><TD>0x00</TD><TD>byte </TD><TD>magic[4]        </TD>
 *  <TD> 0x2E736E64 '.snd' in ASCII     </TD></TR>
 *  <TR><TD>0x04</TD><TD>int </TD><TD>offset </TD>
 *  <TD> offset of audio data relative to the start of the stream </TD></TR>
 *  <TR><TD>0x08</TD><TD>int </TD><TD>size     </TD>
 *  <TD> number of bytes of data        </TD></TR>
 *  <TR><TD>0x0C</TD><TD>int </TD><TD>format   </TD>
 *  <TD> format code: 1 for 8-bit u-law </TD></TR>
 *  <TR><TD>0x10</TD><TD>int </TD><TD>samplingRate </TD>
 *  <TD> the sampling rate              </TD></TR>
 *  <TR><TD>0x14</TD><TD>int </TD><TD>numChannels </TD>
 *  <TD> the number of channels         </TD></TR>
 *  <TR><TD>0x18</TD><TD>byte</TD><TD>info[]      </TD>
 *  <TD> optional text information      </TD></TR>
 *  </TABLE>
 *  </CENTER>
 *  <p>
 *  The design of this class is based on the web page of
 *  <a href=mailto:donahu@cooper.edu>Billy Donahue</a>.
 *  <a href=http://www.cooper.edu/~donahu/auformat/auFormat.html>
 *  http://www.cooper.edu/~donahu/auformat/auFormat.html</a>.
 *  Note that this class serves the same role as
 *  the class by the same name in the sun.audio package, but is much
 *  more public about its information.  For example, the Sun version
 *  does not give any access to the audio data itself.
 *
 * @author Edward A. Lee
 * @version $Id$
 */

public class Audio {

    /** Construct an instance initialized with the audio
     *  signal given by the argument.  The argument is an array
     *  of bytes that are mu-law encoded.  The audio signal is assumed
     *  to have an 8kHz sample rate, with a single channel.
     *  @param audio An audio signal.
     */
    public Audio(byte[] audio) {
        String ptinfo = "Ptolemy audio";
        // NOTE: This uses the platform encoding, which is probably wrong.
        info = ptinfo.getBytes();
        offset = 24 + info.length;
        size = audio.length;
        format = 1;
        sampleRate = 8000;
        numChannels = 1;
        this.audio = new byte[1][];
        this.audio[0] = audio;
    }

    /** Construct an instance initialized with the audio
     *  signal given by the argument.  The argument is an array
     *  of double-precision, floating point audio samples assumed to
     *  be normalized to be in the range -1.0 to 1.0.
     *  The data will be encoded according to the mu-law standard at an
     *  8kHz sample rate, with a single channel.
     *  @param audio An audio signal.
     */
  public Audio(double[] audio) {
        String ptinfo = "Ptolemy audio";
        // NOTE: This uses the platform encoding, which is probably wrong.
        info = ptinfo.getBytes();
        offset = 24 + info.length;
        size = audio.length;
        format = 1;
        sampleRate = 8000;
        numChannels = 1;
        this.audio = new byte[1][size];
        for (int i = size-1; i >= 0; i--) {
            this.audio[0][i] = lin2mu((int)(audio[i]*31616.0));
        }
    }

    /** Construct an instance and initialize it by reading
     *  the specified stream.
     *  @param input The input stream.
     *  @exception IOException If an error occurs reading the input data
     *   (e.g. a premature end of file).
     */
    public Audio(DataInputStream input) throws IOException {
        input.read(magic, 0, 4);

        // Check the magic number, which should be 0x2E736E64, '.snd'
        // in ASCII.
        if (magic[0] != 0x2E || magic[1] != 0x73 || magic[2] != 0x6E ||
                 magic[3] != 0x64) {
             throw new IllegalArgumentException(
                     "ptolemy.media.Audio: bad magic number in "
                     + "stream header.  Not an audio file?");
        }

        offset = input.readInt();
        size = input.readInt();
        format = input.readInt();
        sampleRate = input.readInt();
        numChannels = input.readInt();

        if (offset < 0 || offset > 10000) {
            throw new IllegalArgumentException("ptolemy.media.Audio:"
                    + " offset value '" + offset +
                    "' is out of range 0-10000");
        }
        info = new byte[offset-24];
        input.read(info, 0, offset-24);

        if (format != 1) {
            throw new IllegalArgumentException("ptolemy.media.Audio:"
                    + " Sorry, only 8-bit mu-law encoded data can be read.");
        }
        if (numChannels != 1) {
            throw new IllegalArgumentException("ptolemy.media.Audio:"
                    + " Sorry, only one-channel audio data can be read.");
        }

        // Finally read the audio data.
        audio = new byte[1][size];
        input.readFully(audio[0]);
    }

    ///////////////////////////////////////////////////////////////////
    ////                      public members                         ////

    /** The file type identifier, 0x2E736E64 or '.snd' in ASCII. */
    public byte magic[] = {(byte)'.', (byte)'s', (byte)'n', (byte)'d'};

    /** Offset of audio data relative to the start of the stream. */
    public int offset;

    /** Number of bytes of audio data. */
    public int size;

    /** Format code; 1 for 8-bit mu-law. */
    public int format;

    /** The sampling rate. */
    public int sampleRate;

    /** The number of channels. */
    public int numChannels;

    /** Four byte info field. */
    public byte info[];

    /** Audio data, by channel. */
    public byte[][] audio;

    ///////////////////////////////////////////////////////////////////
    ////                         public methods                    ////

    /** Convert an integer linear representation of an audio sample
     *  into a mu-255 companded representation.  Mu law is the standard
     *  used in Sun .au files as well as throughout the telephone network.
     *  <p>
     *  The integer argument is a 16-bit representation of the sample.
     *  Anything outside the range -32635 to 32635 will be clipped to within
     *  that range.
     *  <p>
     *  The mu-255 representation is a byte SEEEMMMM where S is the sign
     *  bit, EEE is the three-bit exponent, and MMMM is the four-bit
     *  mantissa. The bits are flipped, so that the binary 10000000
     *  is the largest positive number and 00000000 is the largest negative
     *  number. If you have called that static method setZeroTrap() with a
     *  <i>true</i> argument, then
     *  per MIL-STD-188-113, the 00000000 representation
     *  is never used, replaced instead with 00000010 (0x02).
     *  By default, this trap is not used.
     *  <p>
     *  This implementation was written by Anthony Hursh, who included with it
     *  the following information:
     *  <p>
     *  Copyright 1997 by Anthony Hursh
     *  &lt;hursha@saturn.math.uaa.alaska.edu&gt;
     *  This code may be freely used as long as proper credit
     *  is given.  It was originally written in C by
     *  Craig Reese (IDA/Supercomputing Research Center) and
     *  Joe Campbell (Department of Defense), and
     *  ported to Java by Tony Hursh, January 1997.
     *  References:
     *  <ol>
     *  <li> CCITT Recommendation G.711  (very difficult to follow)
     *  <li> "A New Digital Technique for Implementation of Any
     *     Continuous PCM Companding Law," Villeret, Michel,
     *     et al. 1973 IEEE Int. Conf. on Communications, Vol 1,
     *     1973, pg. 11.12-11.17
     *  <li> MIL-STD-188-113,"Interoperability and Performance Standards
     *     for Analog-to_Digital Conversion Techniques,"
     *     17 February 1987
     *  </ol>
     *
     *  @param lin A linear representation of the sample.
     *  @return A mu-255 representation of the sample.
     */
    public static byte lin2mu(int sample) {
        int sign = 0;
        if(sample < 0) {
            sample = -sample;
            sign = 0x80;
        }
        // clip the magnitude
        if (sample > CLIP) sample = CLIP;
        sample = sample + BIAS;
        int exponent =  exp_lut[(sample>>7) & 0xFF];
        int mantissa = (sample >> (exponent+3)) & 0x0F;
        int ulawbyte = (sign | (exponent << 4) | mantissa);
        // System.out.println(" sign = " + sign + " exponent = " +
        // exponent + " mantissa = " + mantissa );
        ulawbyte =  ~ulawbyte;
        ulawbyte &= 0xFF;
        if(_zerotrap && ulawbyte == 0 ) {
            // optional CCITT trap
            ulawbyte = 0x02;
        }
        return (byte)ulawbyte;
    }

    /** Convert mu-255 companded representation of an audio sample
     *  into an integer linear representation.  Mu law is the standard
     *  used in Sun .au files as well as throughout the telephone network.
     *  This implementation is based on the web page by
     *  <a href=mailto:donahu@cooper.edu>Billy Donahue</a>:
     *  <a href=http://www.cooper.edu/~donahu/auformat/auFormat.html>
     *  http://www.cooper.edu/~donahu/auformat/auFormat.html</a>.
     *  The resulting integer values are scaled to be in the range -31616
     *  to 31616.  This uses the low order 16 bits of the resulting integer,
     *  and thus provides a convenient 16-bit linear encoding.
     *  <p>
     *  The mu-255 representation is a byte SEEEMMMM where S is the sign
     *  bit, EEE is the three-bit exponent, and MMMM is the four-bit
     *  mantissa.  The bits are flipped, so that the binary 10000000
     *  is the largest positive number and 00000000 is the largest negative
     *  number.
     *  <p>
     *  If you have called setZeroTrap() with a <i>true</i> argument, then
     *  this will not be an exact inverse of lin2mu because the zero code
     *  is interpreted as being the largest negative number, -31616.
     *
     *  @param b A mu-255 representation of the sample.
     *  @return A linear representation of the sample.
     */
    public static int mu2lin(byte b) {
        // flip the bits
        int mu = b ^ 0xFF;
        int sign = (mu & 0x80) >> 7 ;
        int exponent = (mu & 0x70) >> 4 ;
        int mantissa = (mu & 0x0F);
        // System.out.println(" sign = " + sign + " exponent = " +
        // exponent + " mantissa = " + mantissa );
        int linear = (mantissa<<(exponent+1)) - 0x20 + (0x20 << exponent);
        // Make into a 16 bit sample.
        linear <<= 2;
        return (sign == 1) ? -linear : linear;
    }

    /** Read Sun audio file (.au) format and return the audio data as an array.
     *  The argument stream may represent a file, a URL, or a byte array that
     *  contains the .au format.
     *  For example, given a URL called "url," you can read the audio file
     *  as follows:
     *  <pre>
     *     double audio[] = readAudio(new DataInputStream(url.openStream());
     *  </pre>
     *  <p>
     *  The returned values lie in the range -1.0 to 1.0.
     *
     *  @param input The input stream.
     *  @exception IOException If an I/O error occurs reading the stream.
     *  @exception IllegalArgumentException If the input stream
     *   is not an understood format.  This is a runtime exception, so it need
     *   not be declared explicitly
     */
    public static double[] readAudio(DataInputStream input)
            throws IOException {
        Audio audio = new Audio(input);
        return audio.toDouble(0);
    }

    /** Configure all instances of this class to use the MIL-STD zero trap.
     *  I.e., per MIL-STD-188-113, the 00000000 representation
     *  is never used, replaced instead with 00000010 (0x02).
     *  For some reason, an all zero
     *  mu-law code is sometimes undesirable.  By default, this class does
     *  not use this trap, so you must call this with a <i>true</i> argument
     *  to use the trap.
     *  @param boole If true, use zero-trap encoding.
     */
    public static void setZeroTrap(boolean boole) {
        _zerotrap = boole;
    }

    /** Convert the audio data to linear double encoding (from mu-law).
     *  The returned numbers lie in the range -1.0 to 1.0.
     *  @param channel The channel number.
     *  @return A new array of integers, or null if there is no audio data.
     */
    public double[] toDouble(int channel) {
        int[] intdata = toLinear(channel);
        if (intdata != null) {
            double[] result = new double[intdata.length];
            for (int i = intdata.length-1; i >= 0; i--) {
                result[i] = ((double)(intdata[i]))/31616.0;
            }
            return result;
        }
        return null;
    }

    /** Convert the audio data to linear integer encoding (from mu-law).
     *  The returned integers use the low-order 16 bits only, lying
     *  in the range -31616 to 31616.
     *  @param channel The channel number.
     *  @return A new array of integers, or null if there is no audio data.
     */
    public int[] toLinear(int channel) {
        if (audio != null) {
            if (audio.length > channel && audio[channel] != null) {
                int[] result = new int[audio[channel].length];
                for (int i = audio[channel].length-1; i >= 0; i--) {
                    result[i] = mu2lin(audio[channel][i]);
                }
                return result;
            }
        }
        return null;
    }

    /** Return a readable representation of the header data. */
    public String toString() {
        return "file ID tag = " + new String(magic) + "\n"
            + "offset = " + offset + "\n"
            + "size = " + size + "\n"
            + "format code = " + format + "\n"
            + "sampleRate = " + sampleRate + "\n"
            + "number of channels = " + numChannels + "\n"
            + "info field = " + new String(info).trim();
    }

    /** Write the audio data to an output stream in the Sun audio format.
     *
     *  @param output The output stream.
     *  @exception IOException If an error occurs writing to the stream.
     */
    public void write(DataOutputStream output) throws IOException {
        output.write(magic, 0, 4);
        output.writeInt(offset);
        output.writeInt(size);
        output.writeInt(format);
        output.writeInt(sampleRate);
        output.writeInt(numChannels);
        output.write(info, 0, offset-24);
        output.write(audio[0], 0, size);
    }

    /** Write the raw audio data to an output stream.
     *  This method can be used to play the audio data using the
     *  (undocumented and unsupported) sun.audio package as follows:
     *  <pre>
     *      // The constructor argument below is optional
     *      ByteArrayOutputStream out =
     *               new ByteArrayOutputStream(sound.size);
     *      try {
     *          sound.writeRaw(new DataOutputStream(out));
     *      } catch (IOException ex) {
     *          throw new RuntimeException("Audio output failed");
     *      }
     *      byte[] iobuffer = out.toByteArray();
     *      ByteArrayInputStream instream =
     *              new ByteArrayInputStream(_iobuffer);
     *      AudioPlayer.player.start(instream);
     *  </pre>
     *  The above code assumes we have an sun.audio.AudioData object
     *  called "sound".
     *  Although it would seem reasonable to include a "play" method in this
     *  class to do this, we wish to avoid a dependence on the sun.audio
     *  package in this Ptolemy package, so you will have to implement
     *  the above code yourself.
     *
     *  @param output The output stream.
     *  @exception IOException If an error occurs writing to the stream.
     */
    public void writeRaw(DataOutputStream output) throws IOException {
        output.write(audio[0], 0, size);
    }

    /** Write Sun audio file (.au) format from an array.
     *  The argument is an array of doubles assumed to lie in the range of
     *  -1.0 to 1.0.  The data is converted to one-channel mu-law samples
     *  at 8kHz.
     *
     *  @param audio The audio data, as an array of doubles.
     *  @param output The output stream.
     *  @xception IOException If an I/O error occurs writing to the stream.
     */
    public static void writeAudio(double[] audio, DataOutputStream output)
            throws IOException {
        Audio obj = new Audio(audio);
        obj.write(output);
    }

    ///////////////////////////////////////////////////////////////////
    ////                         private members                         ////

    // The following are used for mu-law conversion.
    // Turn on the trap as per the MIL-STD (this prevents a result of 0).
    private static boolean _zerotrap = false;

    // define the add-in bias for 16 bit samples.
    private static final int BIAS = 0x84;

    // clipping value for inputs.
    private static final int CLIP =  32635;

    // lookup table for the exponent.
    private static final byte exp_lut[] = {0, 0, 1, 1, 2, 2, 2, 2, 3,
3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7}; }

File: ptolemy/actor/lib/jjs/modules/audio/audio.js
// Below is the copyright agreement for the Ptolemy II system.
//
// Copyright (c) 2015-2016 The Regents of the University of California.
// All rights reserved.
//
// Permission is hereby granted, without written agreement and without
// license or royalty fees, to use, copy, modify, and distribute this
// software and its documentation for any purpose, provided that the above
// copyright notice and the following two paragraphs appear in all copies
// of this software.
//
// IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
// FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
// ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
// THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
// SUCH DAMAGE.
//
// THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
// PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
// CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
// ENHANCEMENTS, OR MODIFICATIONS.
//
//
// Ptolemy II includes the work of others, to see those copyrights, follow
// the copyright link on the splash page or see copyright.htm.

/**
 * Module to access audio hardware on the host.
 * This module defines three key classes, Player, ClipPlayer, and
 * Capture.  The Player class accepts audio data in a variety of formats
 * and plays them through the default audio output device on the host
 * platform. The ClipPlayer accepts a URL specifying an audio source
 * file and plays that.  The Capture class records audio from the default
 * audio input device, such as a microphone, on the host platform.
 * 
 * @module audio
 * @author Edward A. Lee and Beth Osyk
 */

// Stop extra messages from jslint.  Note that there should be no
// space between the / and the * and global.
/*globals Java, error, exports */
/*jshint globalstrict: true*/
"use strict";

var AudioHelper = Java.type('ptolemy.actor.lib.jjs.modules.audio.AudioHelper');
var EventEmitter = require('events').EventEmitter;

// Clip playback uses javafx instead of Ptolemy SoundReader since javafx supports mp3
var AudioClip = Java.type('javafx.scene.media.AudioClip');

/** Return an array of byte formats currently available for audio capture
 *  on the current host. This includes at least 'raw', which is whatever
 *  byte format the host audio hardware happens to provide when capturing
 *  audio from the microphone.
 *  @return An array of capture formats.
 */
exports.byteFormats = function () {
    // The Java.from() Nashorn extension converts a Java array into a JavaScript array.
    return Java.from(AudioHelper.byteFormats());
};

/** Construct an instance of an Player object type. This should be instantiated in your
 *  JavaScript code as
 *  <pre>
 *     var audio = require("audio");
 *     var player = new audio.Player();
 *  </pre>
 *  An instance of this object type implements the following functions:
 *  <ul>
 *  <li> play(data, callback): Play the specified array, and call the specified
 *       callback function when the data has been accepted.
 *  <li> stop(): Stop playback and free the audio resources.
 *  </ul>
 *  
 *  The callback function should be used to construct the next batch of
 *  audio data to be played. The callback function will be called as soon
 *  as the specified data has been queued to the audio system.
 *  You should not call the play() function again until the callback has
 *  been called or you might overwrite the audio data.
 *  
 *  The _playbackFormat_ argument specifies the form in which the audio
 *  data will be provided. The available formats include:
 *  
 *  * "raw": The data is a byte array representing audio data exactly as
 *    captured by default on the host.
 *  * "array": The audio input data is an array of arrays of numbers,
 *    where each number is in the range from -1.0 to 1.0.
 *    The first index of the input specifies the channel number.
 *  * "encoded": The audio input data is a byte array containing audio
 *    data encoded in one of the file format standards such as 
 *    AIFF (historically associated with Apple computers),
 *    AIFF-C (a compressed version of AIFF),
 *    AU (historically associated with Sun Microsystems and Unix computers), or
 *    WAVE (historically associated with Windows PCs).
 *  
 *  The optional _playbackOptions_ argument is an object with the following properties,
 *  all of which are optional:
 *  
 *  * _bitsPerSample_: The number of bits per sample. This is an integer that
 *    defaults to 16.
 *  * _channels_: The number of channels. This defaults to 1.
 *  * _sampleRate_: The sample rate. This is an integer that defaults to 8000.
 *    Typical supported sample rates are 8000, 11025, 22050, 44100, and 48000.
 *    
 *  The audio data will be converted to this specified format before being sent
 *  to the audio hardware.
 *  
 *  @param playbackFormat The format of the input data.
 *  @param playbackOptions The playback options.
 */
exports.Player = function (playbackFormat, playbackOptions) {
    this.helper = new AudioHelper(actor, this);

    playbackOptions = playbackOptions || {};
    playbackOptions.bitsPerSample = playbackOptions.bitsPerSample || 16;
    playbackOptions.channels = playbackOptions.channels || 1;
    playbackOptions.sampleRate = playbackOptions.sampleRate || 8000;

    this.helper.setPlaybackParameters(playbackOptions, playbackFormat);
    
    this.playbackFormat = playbackFormat;
};

/** Play audio data. This function returns immediately, but this does not
 *  mean that the samples have been played. In fact, they have not even been
 *  queued to audio system, necessarily.
 *  This function should not be called again before the callback has
 *  been invoked.
 *  @param data An array of arrays of numbers in the range -1 to 1 to be played.
 *  @param callback A callback function to invoke when the samples have been
 *   queued to the audio system.
 */
exports.Player.prototype.play = function (data, callback) {
    // Start playback, if it isn't already running.
    this.helper.startPlayback();

    if (this.playbackFormat === 'array') {
        this.helper.putSamples(data, callback);
    } else {
        this.helper.putBytes(data, callback);
    }
};

/** Stop playback. */
exports.Player.prototype.stop = function () {
    this.helper.stopPlayback();
};


/** Construct an instance of a ClipPlayer object type.  A ClipPlayer plays
 *  audio from a URL source. This should be instantiated in your JavaScript code as:
 *  <pre>
 *     var audio = require("audio");
 *     var player = new audio.ClipPlayer(url);
 *  </pre>
 *  An instance of this object type implements the following functions:
 *  <ul>
 *  <li> play(): Play the audio from the specified url.
 *  <li> stop(): Stop playback.
 *  </ul>
 */

/** Create a ClipPlayer to play the specified URL.
 */
exports.ClipPlayer = function (url) {
    try {
        this.clip = new AudioClip(url);
    } catch (err) {
        error("Error connecting to audio URL " + url);
    }
};

/** Play the currently loaded audio clip.
 */
exports.ClipPlayer.prototype.play = function () {
    if (this.clip !== null) {
        this.clip.stop();
        this.clip.play();
    } else {
        error("No audio clip to play.  Please load a url first.");
    }
};

/** Stop playback.
 */
exports.ClipPlayer.prototype.stop = function () {
    if (this.clip !== null) {
        this.clip.stop();
    }
};

/** Construct an instance of a Capture object type. This should be instantiated in your
 *  JavaScript code as
 *  <pre>
 *     var audio = require("audio");
 *     var capture = new audio.Capture();
 *  </pre>
 *  An instance of this object type implements the following functions:
 *  <ul>
 *  <li> capture.start(): Start capturing.
 *  <li> stop(): Stop capturing.
 *  </ul>
 *  This is an event emitter that will emit an event "capture"
 *  when audio data has been captured. The argument passed
 *  to a listener for that event will be audio data in a format
 *  given by the _outputFormat_ argument, which is one of the
 *  following strings:
 *  
 *  * "raw": The argument is a byte array representing audio data exactly as
 *    captured in the specified capture format.
 *  * "array": The audio data is converted into arrays of numbers (one per
 *    channel), where each number is in the range from -1.0 to 1.0.
 *    The argument passed to the "capture" event is an array of arrays,
 *    where the first index specifies the channel number.
 *  * "aiff": The audio data is converted into the AIFF file format historically
 *    associated with Apple computers.
 *  * "aifc": The audio data is converted into the AIFF-C, a compressed version
 *    of AIFF.
 *  * "au": The audio data is converted into the AU file format historically
 *    associated with Sun Microsystems and Unix computers.
 *  * "wav": The audio data is converted into the WAVE file format historically
 *    associated with Windows PCs.
 *    
 *  The optional _captureOptions_ argument is an object with the following properties,
 *  all of which are optional:
 *  
 *  * _bitsPerSample_: The number of bits per sample. This is an integer that
 *    defaults to 16.
 *  * _channels_: The number of channels. This defaults to 1.
 *  * _sampleRate_: The sample rate. This is an integer that defaults to 8000.
 *  
 *  @param captureTime The length of time for each audio capture (in milliseconds).
 *   This is an integer that defaults to 1000, capturing 1 second of audio at a time.
 *  @param outputFormat The format of the data passed to the "capture" event
 *   listeners as specified above.
 */
exports.Capture = function (captureTime, outputFormat, captureOptions) {
    this.helper = new AudioHelper(actor, this);

    captureOptions = captureOptions || {};
    captureOptions.bitsPerSample = captureOptions.bitsPerSample || 16;
    captureOptions.channels = captureOptions.channels || 1;
    captureOptions.sampleRate = captureOptions.sampleRate || 8000;

    this.helper.setCaptureParameters(captureOptions, captureTime, outputFormat);
    
    this.outputFormat = outputFormat;
};
util.inherits(exports.Capture, EventEmitter);

/** Start the capture. */
exports.Capture.prototype.start = function () {
    this.helper.startCapture();
};

/** Stop the capture and free audio resources. */
exports.Capture.prototype.stop = function () {
    this.helper.stopCapture();
};

/** Callback function used by the helper to deliver data.
 *  @param audioData The audio data.
 */
exports.Capture.prototype._captureData = function(audioData) {
    if (this.outputFormat == 'array') {
        // Use Nashorn-specific conversion to convert to a JavaScript array.
        var channels = Java.from(audioData);
        for (var i = 0; i < channels.length; i++) {
            channels[i] = Java.from(channels[i]);
        }
        this.emit('capture', channels);
    } else if (this.outputFormat == 'samples') {
        // NOTE: This is not documented nor encouraged anymore.
        if (audioData.length === 1) {
            for (var i = 0; i < audioData[0].length; i++) {
                this.emit('capture', audioData[0][i]);
            }
        } else {
            // Assume all channels have the same length.
            for (var i = 0; i < audioData[0].length; i++) {
                var output = [];
                for (var j = 0; j < audioData.length; j++) {
                    output[j] = audioData[j][i];
                }
                this.emit('capture', output);
            }
        }
    } else {
        // Emit a byte array.
        this.emit('capture', Java.from(audioData));
    }
}
